pkg/controller/kubelet-config/kubelet_config_bootstrap.go

```diff
@@ -0,0 +1,175 @@
+package kubeletconfig
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	"github.com/openshift/machine-config-operator/pkg/version"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+const (
+	kubeletCfgMasterLabel = "pools.operator.machineconfiguration.openshift.io/master"
+	kubeletCfgWorkerLabel = "pools.operator.machineconfiguration.openshift.io/worker"
+)
+
+// RunKubeletBootstrap generates MachineConfig objects for mcpPools that would have been generated by syncKubeletConfig
+func RunKubeletBootstrap(templateDir string, kubeletConfigs []*mcfgv1.KubeletConfig, controllerConfig *mcfgv1.ControllerConfig, features *configv1.FeatureGate, mcpPools []*mcfgv1.MachineConfigPool) ([]*mcfgv1.MachineConfig, error) {
+	generatedManagedKeys := make(map[string]string)
+	generatedMCs := make(map[string]*mcfgv1.MachineConfig)
+	// Validate the KubeletConfig CR if exists
+	for _, kubeletConfig := range kubeletConfigs {
+		if kubeletConfig == nil {
+			continue
+		}
+		if err := validateUserKubeletConfig(kubeletConfig); err != nil {
+			return nil, newForgetError(err)
+		}
+	}
+
+	userDefinedSystemReserved := make(map[string]string)
+	var res []*mcfgv1.MachineConfig
+	for _, kubeletConfig := range kubeletConfigs {
+		if kubeletConfig == nil || kubeletConfig.Spec == (mcfgv1.KubeletConfigSpec{}) {
```

Comment:
this isn't the correct way to check for a type. is this really checking for a type?

Reply:
No. It's checking if the Spec is empty struct.

Reply:
Is this new code or was it somewhere else in the code base?

Reply:
this is new code.

Reply:
oh, the new code added since comments worried about the spec might be empty https://github.com/openshift/machine-config-operator/pull/2547#discussion_r667264496
We don't have to worry about it's empty since it's required? https://github.com/openshift/machine-config-operator/blob/master/pkg/apis/machineconfiguration.openshift.io/v1/types.go#L346-L347